### PR TITLE
Privacy token

### PR DESCRIPTION
**Pull request by anhntbk08**

### CODE REVIEW

File: contracts/PToken.sol [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+                _data);
+    }
+
+    function transferFee(uint256 fee) internal {
+        uint256 _externalValue = _toExternalValue(fee);
+        ITRC21 _token = ITRC21(token());
+        _token.transfer(issuer(), _externalValue);
+
+        bool success;
+        assembly {
```
Comment (phamvancam2104) [inline contracts/PToken.sol]: why do we need this code portion?

File: contracts/PToken.sol [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+                _mask,
+                _amount,
+                _encodedMask,
+                _data);
+    }
+
+    function transferFee(uint256 fee) internal {
+        uint256 _externalValue = _toExternalValue(fee);
+        ITRC21 _token = ITRC21(token());
+        _token.transfer(issuer(), _externalValue);
```
Comment (phamvancam2104) [inline contracts/PToken.sol]: this needs to be very careful as the actual token transferred from the contract would be: the fee of the token (caused by `transfer` function) + `_externalValue`, which results in the sum of the token hold by the contract is less than the sum of all privacy token holders => withdrawals will fail

Comment (anhntbk08) [reply contracts/PToken.sol]: :+1:

File: contracts/PTRC21.sol [hunk 3]
```diff
@@ -9,18 +10,148 @@ import "./CopyUtils.sol";
...
+    function symbol() public view returns (string memory);
+    function decimals() public view returns (uint8);
+}
+
+contract PTRC21 is TRC21 {
+    string private _name;
+    string private _symbol;
+    uint8 private _decimals;
+    uint8 constant PRIVACY_DECIMALS = 8;
+    uint256 private _minFee;
```
Comment (phamvancam2104) [inline contracts/PTRC21.sol]: Remove `_minFee` as it is in TRC21.sol already

File: contracts/PTRC21.sol [hunk 4]
```diff
@@ -9,18 +10,148 @@ import "./CopyUtils.sol";
...
 
+contract AbstractTokenTRC21 {
+    function issuer() public view returns (address);
+    function name() public view returns (string memory);
+    function symbol() public view returns (string memory);
+    function decimals() public view returns (uint8);
+}
+
+contract PTRC21 is TRC21 {
+    string private _name;
```
Comment (phamvancam2104) [inline contracts/PTRC21.sol]: all of those fields could have `public` access

File: contracts/PTRC21.sol [hunk 5]
```diff
@@ -9,18 +10,148 @@ import "./CopyUtils.sol";
...
+    address private _token;
+
+    constructor (address token,
+        uint256 minFee,
+        uint256 depositFee,
+        uint256 withdrawFee
+    ) public {
+        if (token != address(0)) {
+            AbstractTokenTRC21 t = AbstractTokenTRC21(token);
+            // require(t.issuer() == msg.sender);
```
Comment (phamvancam2104) [inline contracts/PTRC21.sol]: require or not?

File: contracts/PTRC21.sol [hunk 6]
```diff
@@ -9,18 +10,148 @@ import "./CopyUtils.sol";
...
+
+            _token = token;
+            _name =  string(abi.encodePacked("P", t.name()));
+
+            _symbol = t.symbol();
+            _decimals = t.decimals();
+        } else {
+            // Tomo token
+            _token = address(0);
+            _name = "TomoP";
```
Comment (phamvancam2104) [inline contracts/PTRC21.sol]: name = `TomoP Token`?
Symbol = `TPT`?

File: contracts/PTRC21.sol [hunk 7]
```diff
@@ -9,18 +10,148 @@ import "./CopyUtils.sol";
...
+            _symbol = "TomoP";
+            _decimals = 18;
+        }
+
+        // change the fee correspondent to new decimal in the privacy system
+        (_minFee, _depositFee, _withdrawFee) = _calculatePrivacyFee(minFee, depositFee, withdrawFee);
+        _changeMinFee(minFee);
```
Comment (phamvancam2104) [inline contracts/PTRC21.sol]: if minFee > 0, then privacy contracts wont work as consensus will check that, all transaction senders (which were randomly generated) need to have at least `minFee` amount of token => review this again.

File: contracts/PTRC21.sol [hunk 8]
```diff
@@ -9,18 +10,148 @@ import "./CopyUtils.sol";
...
+    }
+
+    /**
+     * @return reduce decimal length.
+     */
+    function _calculatePrivacyFee(uint256 minFee, uint256 depositFee, uint256 withdrawFee) internal view returns (uint256, uint256, uint256) {
+        return (_toPrivacyValue(minFee), _toPrivacyValue(depositFee), _toPrivacyValue(withdrawFee));
+    }
+
+    // convert external value to privacy's
```
Comment (phamvancam2104) [inline contracts/PTRC21.sol]: a bit redundant? as 2 functions basically do the same thing, which can be replaced by a function: convertUnits ...

File: contracts/PTRC21.sol [hunk 9]
```diff
@@ -9,18 +10,148 @@ import "./CopyUtils.sol";
...
+    /**
+     * @return the number of decimals of the token.
+     */
+    function privacyDecimals() public pure returns (uint8) {
+        return PRIVACY_DECIMALS;
+    }
+
+    function setMinFee(uint256 value) public {
+        require(msg.sender == issuer());
+        _minFee = _toPrivacyValue(value);
```
Comment (phamvancam2104) [inline contracts/PTRC21.sol]: `_minFee`of PTRC21 or _minFee of TRC21?

File: contracts/PTRC21.sol (hunk 8 above)
Comment (anhntbk08) [reply contracts/PTRC21.sol]: _calculatePrivacyFee will be removed

File: contracts/PToken.sol (hunk 1 above)
Comment (anhntbk08) [reply contracts/PToken.sol]: Will be removed.

**Pull request merged**